### ISSUE TITLE
Can we report the source?

### ISSUE THREAD

**Issue by edeutsch** [NCATSTranslator/reasoner-validator#86]
A common problem is that ARAX output is declared invalid (i.e. has errors), e.g.:

https://arax.ncats.io/devED/?r=146888
```
Errors:
* Knowledge Graph Edge Attribute Type Id Not:
=> Edge has a value that is not a CURIE for attribute_type_id
	# article_doi:
	- edge_id: 
		CHEMBL.COMPOUND:CHEMBL112--biolink:physically_interacts_with->NCBIGene:10257
...
	# bindingdb_curation_datasource:
	- edge_id: 
		CHEMBL.COMPOUND:CHEMBL112--biolink:physically_interacts_with->NCBIGene:10257
...
	# in_taxon:
	- edge_id: 
		CHEMBL.COMPOUND:CHEMBL112--biolink:physically_interacts_with->NCBIGene:10257
...
```

And it takes some careful sleuthing to figure that infact ARAX is not at fault here, it is BTE that is providing the invalid data to ARAX, which is just carrying it along.

Would it not-too-difficult is show the sources for these problems?
e.g. for the above problems, it would be nice to see:
```
sources: infores:bindingdb --> infores:biothings-bindingdb --> infores:biothings-explorer --> infores:arax
```

This would allow me to trivially see that this "ARAX error" is really a BTE KP error that ARAX is just carrying along.

that one-liner would be translated from:
```
          "sources": [
            {
              "resource_id": "infores:bindingdb",
              "resource_role": "primary_knowledge_source",
              "source_record_urls": [],
              "upstream_resource_ids": []
            },
            {
              "resource_id": "infores:biothings-bindingdb",
              "resource_role": "aggregator_knowledge_source",
              "source_record_urls": [],
              "upstream_resource_ids": [
                "infores:bindingdb"
              ]
            },
            {
              "resource_id": "infores:biothings-explorer",
              "resource_role": "aggregator_knowledge_source",
              "source_record_urls": [],
              "upstream_resource_ids": [
                "infores:biothings-bindingdb"
              ]
            },
            {
              "resource_id": "infores:arax",
              "resource_role": "aggregator_knowledge_source",
              "source_record_urls": [],
              "upstream_resource_ids": [
                "infores:biothings-explorer"
              ]
            }
          ],
```

What do you think?

**Comment by RichardBruskiewich** [comment]
OK, I'll take a run at this very quickly

**Comment by RichardBruskiewich** [comment]
Hi again @edeutsch, it took me a couple of days to get to this issue, but working on it right now. I might hopefully have the code completed and released by the end of the day, time permitting.

I'll point out the obvious: that this reporting can't apply to the TRAPI schema validation itself - since that is done in an opaque 'whole JSON message' lack box manner using JSONSchema - but would only apply from the validation of specific instances of **`Message.knowledge_graphs.Edge`** data.

**Comment by edeutsch** [comment]
Agreed and understood. It is mostly relevant for the problems with attributes.

**Comment by RichardBruskiewich** [comment]
Moving right along @edeutsch, I note that `**upstream_resource_ids**` is a list, which suggests a directed acyclic graph for provenances. Not sure what that means in this context. What was the use case for this?

**Comment by edeutsch** [comment]
Mostly for when merging of results takes place. If results from multiple ARAs is merged, there could be a complicated DAG. But for now, just  using the first if there are multiple is fine. Maybe add an asterisk to signal ignored complications.

**Comment by RichardBruskiewich** [comment]
Roger that!

**Comment by RichardBruskiewich** [comment]
Hmm.. I guess **`upstream_resource_ids`** means "from the **`primary_knowledge_source`** towards the topmost ARA or KP exporter (aggregator) for the edge. Your requested provenance path string has this semantic:

```
sources: infores:bindingdb --> infores:biothings-bindingdb --> infores:biothings-explorer --> infores:arax
```

I didn't initially pick up on that (I guess I wasn't paying enough attention to the TRAPI working group discussions about this...)

**Comment by edeutsch** [comment]
I'm uncertain if this is a question, but if so, I think the answer may be yes!

**Comment by RichardBruskiewich** [comment]
Aiming for the following output for sources:
```
Errors:
* Knowledge Graph Edge Attribute Type Id Not:
=> Edge has a value that is not a CURIE for attribute_type_id
       $ infores:bindingdb --> infores:biothings-bindingdb --> infores:biothings-explorer --> infores:arax
	    # article_doi:
	    - edge_id: 
		    CHEMBL.COMPOUND:CHEMBL112--biolink:physically_interacts_with->NCBIGene:10257
...
      $ infores:bindingdb --> infores:biothings-bindingdb --> infores:biothings-explorer --> infores:arax
	    # bindingdb_curation_datasource:
	    - edge_id: 
		    CHEMBL.COMPOUND:CHEMBL112--biolink:physically_interacts_with->NCBIGene:10257
...
      $ infores:bindingdb --> infores:biothings-bindingdb --> infores:biothings-explorer --> infores:arax
	    # in_taxon:
	    - edge_id: 
		     CHEMBL.COMPOUND:CHEMBL112--biolink:physically_interacts_with->NCBIGene:10257
...
```

where the $ sign will be the "sources" scope line, which will have one of two values: 'global' or the 'resources_id' audit trail string as you provided above.

**Comment by edeutsch** [comment]
Yes, this would be great.

btw, it would be great also to have a "..." programmatically, too. i.e. maybe show 3 edge_ids and then if there are more than 3, have a:

... (25 more)

line. This would make things more compact without losing much.
Would require a little fancy footwork in the program of course..

**Comment by RichardBruskiewich** [comment]
resolved by release 3.7.0
